[BOLT] Use symbol table info in registerFragment

aaupov opened the pull request:
Fragment matching relies on symbol names to identify and register split
function fragments. However, as split fragments are often local symbols,
name aliasing is possible. For such cases, use symbol table to resolve
ambiguities.

This requires the presence of FILE symbols in the input binary. As BOLT
requires non-stripped binary, this is a reasonable assumption. Note that
`strip -g` removes FILE symbols by default, but `--keep-file-symbols`
can be used to preserve them.

Depends on: https://github.com/llvm/llvm-project/pull/89861

Test Plan:
Updated X86/fragment-lite.s

aaupov committed:
[BOLT] Use symbol table info in registerFragment

Fragment matching relies on symbol names to identify and register split
function fragments. However, as split fragments are local symbols,
name aliasing is possible. For such cases, use symbol table to resolve
ambiguities:
- find FILE symbol the fragment belongs to,
- iterate over local symbols belonging to the same file and check if
  parent symbol is present,
- if a local parent symbol is found, use it, otherwise use global
  parent symbol.

This requires the presence of FILE symbols in the input binary. As BOLT
requires non-stripped binary, this is a reasonable assumption. Note
however that `strip -g` removes FILE symbols by default, but
`--keep-file-symbols` can be used to preserve these.

We opted to resolve ambiguous cases one by one because there are
typically very few such cases and symbol table parsing is relatively
expensive:

python3.8.6:
Total symbols: 17522
Function symbols: 7877 (45% total)
Fragments: 2551 (15% total)
Fragments with ambiguous parent: 31 (0.1% total)

Large binary:
Total symbols: 3032180
Function symbols: 1637209 (54% total)
Fragments: 43077 (0.1% total)
Fragments with ambiguous parent: 180 (0.006% total)

Test Plan:
Updated X86/fragment-lite.s

aaupov committed:
fixup! [BOLT] Use symbol table info in registerFragment

dcci commented on bolt/include/bolt/Rewrite/RewriteInstance.h:
```diff
@@ -490,6 +490,9 @@ class RewriteInstance {
   std::map<uint64_t, llvm::object::SymbolRef> FileSymRefs;
 
+  /// FILE symbols.
+  std::vector<object::DataRefImpl> FileSymbols;
```
This comment is perfunctory. You might want to consider making it more explicit or just remove it altogether.

dcci commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
+        AmbiguousFragments.emplace_back(ParentName, &Function);
         continue;
       }
+      assert(BD);
```
add a message?

dcci replied:
Also, why is this an assertion and not an `exit(1)`-style error?

aaupov replied:
I'm on the fence regarding making it an error over an assertion.
Assertions should be used for internal invariants, whereas exit(1)/errors should be used for validating input. The absence of a parent fragment is the latter and that's why I used exit(1) in BinaryFunction lookup. I guess I should treat BD similarly because BD->getAddress would crash in Release build mode.

dcci commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
+    // No local parent is found, use global parent function.
+    if (!ParentAddress) {
+      BinaryData *ParentBD = BC->getBinaryDataByName(ParentName);
+      assert(ParentBD);
```
ditto

dcci commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,106 @@ void RewriteInstance::registerFragments() {
+    }
+
+    BinaryFunction *ParentBF = BC->getBinaryFunctionAtAddress(ParentAddress);
+    assert(ParentBF);
```
ditto

aaupov committed:
Also check function aliases

aaupov committed:
Switch to exit(1) in BD/BF checks

aaupov committed:
fixup! Also check function aliases

aaupov committed:
reduce changes

aaupov committed:
Add a comment with reasoning behind handling ambiguous fragments one by one

maksfb commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,108 @@ void RewriteInstance::registerFragments() {
+      uint64_t NumPossibleLocalParents = NR.getNumDuplicates(ParentName);
+      // The most common case: single local parent fragment.
+      if (!BD && NumPossibleLocalParents == 1) {
+        BD = BC->getBinaryDataByName((Twine(ParentName) + "/1").str());
```
Please use `NameResolver` interface.

aaupov commented on bolt/include/bolt/Rewrite/RewriteInstance.h:
```diff
@@ -490,6 +490,9 @@ class RewriteInstance {
   std::map<uint64_t, llvm::object::SymbolRef> FileSymRefs;
 
+  /// FILE symbols used for disambiguating split function parents.
+  std::vector<object::DataRefImpl> FileSymbols;
```
Yes. SymbolRef contains DataRefImpl and an object pointer which in our case is always `InputFile`.

maksfb replied:
If memory is a concern, you should free it once done processing. I find the savings questionable to be worth it over readability, especially when you compare symbols.

aaupov commented on bolt/include/bolt/Utils/NameResolver.h:
```diff
@@ -28,6 +28,13 @@ class NameResolver {
   static constexpr char Sep = '/';
 
 public:
+  /// Return the number of duplicates registered for a given \p Name.
```
Let me rename/rephrase it to "counter" instead to make it more clear.
I took "duplicates" from the comment above.

aaupov committed:
Address comments

aaupov committed:
Address comments

maksfb commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
-        continue;
+      if (BD) {
+        const uint64_t Address = BD->getAddress();
+        BinaryFunction *BF = BC->getBinaryFunctionAtAddress(Address);
```
```suggestion
        BinaryFunction *BF = BC->getFunctionForSymbol(BD->getSymbol());
```

maksfb commented on bolt/include/bolt/Utils/NameResolver.h:
```diff
@@ -28,10 +28,23 @@ class NameResolver {
 
 public:
+  /// Return the counter for a given \p Name.
+  uint64_t getCounter(StringRef Name) const {
```
`getUniquifiedNameCount()`?

maksfb commented on bolt/include/bolt/Utils/NameResolver.h:
```diff
@@ -28,10 +28,23 @@ class NameResolver {
+  }
+
   /// Return unique version of the \p Name in the form "Name<Sep><Number>".
+  std::string getUniqueName(StringRef Name, const uint64_t ID) const {
```
Update the commend above with `ID`.

aaupov commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
     for (StringRef Name : Function.getNames()) {
       StringRef BaseName, Suffix;
       std::tie(BaseName, Suffix) = Name.split('/');
       const size_t ColdSuffixPos = BaseName.find(".cold");
```
Suffix is used to determine if a given fragment name is a local symbol. How can I check it otherwise?

maksfb replied:
`BaseName == Name`?

aaupov replied:
Oh right, I used a similar check just recently in an [alternative implementation](https://github.com/llvm/llvm-project/pull/89320/files#diff-4be6195ccdb53447f870b9b26a36d0b916ace3cbd6efe595bb3cbffd53cf0dacR1441).

aaupov committed:
Store FileSymbols as SymbolRefs

aaupov committed:
Store FileSymbols as ELFSymbolRefs.

Simplify lookups and comparisons using https://github.com/llvm/llvm-project/pull/89861

aaupov committed:
Clear FileSymbols after registerFragments

aaupov commented:
@maksfb: I've addressed the existing concerns – does that look good for you? Note we now depend on #89861 for ELFSymbolRef comparisons. We can add heuristics based on bolt_pseudo.o to make fragment matching even faster in BOLTed binary case as follow-up diffs.

maksfb reviewed (commented):
Otherwise LGTM.

aaupov committed:
Rename FirstGlobal to LocalSymEnd, remove error exit

aaupov commented on bolt/lib/Rewrite/RewriteInstance.cpp:
```diff
@@ -1417,50 +1418,125 @@ void RewriteInstance::registerFragments() {
+      break;
+    }
+  }
+  if (FirstGlobal == 0) {
```
I've re-read the documentation carefully and it says that sh_info has the value one larger than the last local symbol index: https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-94076.html#scrolltoc (Table 12-9 ELF sh_link and sh_info Interpretation).
So I'm dropping the check and exit.
